`detect_init_system` is broken for the dockerized AGW

### Your Environment

Dockerized AGW setup as described in the documentation.

### Describe the Issue

#14004 broke the `s1ap_utils.detect_init_system` method since it replaced try/except clauses and some errors now are no longer correctly handled.

`detect_init_system` first uses `exec_command_output` and thereby `subprocess.check_output` to call `systemctl is-active magma@magmad` on the dev VM. However, the `systemctl` returns an exit code 3 if the service is not running (e.g. if we use the dockerized AGW setup). `subprocess.check_output` then fails with a `subprocess.CalledProcessError`, which before #14004 was excepted.

#### To Reproduce

Follow the dockerized AGW setup as described in magma/lte/gateway/docker/README.md, start tests with, e.g. `make nonsanity`.

#### Expected behavior

The tests would start running.

#### Screenshots

See the stacktrace below.

```
==================================================================================================================================== FAILURES ====================================================================================================================================
_________________________________________________________________________________________________________ TestModifyConfigForNonSanity.test_modify_config_for_non_sanity _________________________________________________________________________________________________________

self = <test_modify_config_for_non_sanity.TestModifyConfigForNonSanity testMethod=test_modify_config_for_non_sanity>

    def test_modify_config_for_non_sanity(self):
        """Modify configurations for Non-Sanity Test Cases
    
        Some non-sanity test cases need changes in default configuration. This
        test script modifies the  configuration files with required values so
        that all the test cases of non-sanity test suite can be verified
        """
        magmad_client = MagmadServiceGrpc()
>       self._magmad_util = MagmadUtil(magmad_client)

s1aptests/test_modify_config_for_non_sanity.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
s1aptests/s1ap_utils.py:877: in __init__
    self._init_system = self.detect_init_system()
s1aptests/s1ap_utils.py:933: in detect_init_system
    systemd_magmad_running = self.exec_command_output(
s1aptests/s1ap_utils.py:912: in exec_command_output
    return subprocess.check_output(
/usr/lib/python3.8/subprocess.py:415: in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True, popenargs = (['sshpass', '-p', 'vagrant', 'ssh', '-o', 'UserKnownHostsFile=/dev/null', ...],), kwargs = {'shell': False, 'stdout': -1}, process = <subprocess.Popen object at 0x7f5e30ae4190>
stdout = b'inactive\n', stderr = None, retcode = 3

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '['sshpass', '-p', 'vagrant', 'ssh', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'StrictHostKeyChecking=no', '-o', 'LogLevel=ERROR', 'vagrant@192.168.60.142', 'systemctl is-active magma@magmad']' returned non-zero exit status 3.

/usr/lib/python3.8/subprocess.py:516: CalledProcessError
------------------------------------------------------------------------------------------------ generated xml file: /var/tmp/test_results/test_modify_config_for_non_sanity.xml -------------------------------------------------------------------------------------------------
============================================================================================================================ short test summary info =============================================================================================================================
FAILED s1aptests/test_modify_config_for_non_sanity.py::TestModifyConfigForNonSanity::test_modify_config_for_non_sanity - subprocess.CalledProcessError: Command '['sshpass', '-p', 'vagrant', 'ssh', '-o', 'UserKnownHostsFile=/dev/null', '-o', 'StrictHostKeyChecking=no', '-...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=============================================================================================================================== 1 failed in 0.68s ================================================================================================================================
make: *** [Makefile:86: nonsanity] Error 1
```

--- a/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
+++ b/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
@@ -934,8 +934,12 @@ class MagmadUtil(object):
         res_systemd = self.exec_command_capture_output(
             "systemctl is-active magma@magmad",
         )
-        docker_magmad_running = res_docker.stdout.decode("utf-8").strip('\n') == 'magmad'
-        systemd_magmad_running = res_systemd.stdout.decode("utf-8").strip('\n') == 'active'
+        docker_magmad_running = \
+            self._is_installed("docker") & \
+            (res_docker.stdout.decode("utf-8").strip('\n') == 'magmad')
+        systemd_magmad_running = \
+            self._is_installed("systemctl") & \
+            (res_systemd.stdout.decode("utf-8").strip('\n') == 'active')
 
         if systemd_magmad_running:
             # default to systemd if docker and systemd are running - needed by feg integ tests
@@ -947,6 +951,13 @@ class MagmadUtil(object):
                 "Magmad is not running, you have to start magmad "
                 "either in Docker or systemd",
             )
+
+    def _is_installed(self, cmd):
+        """Check if a command is installed on the system."""
+        is_installed = self.exec_command(f"type {cmd} >/dev/null 2>&1") == 0
+        if not is_installed:
+            logging.info(f"{cmd} is not installed")
+        return is_installed
 
     @property
     def init_system(self):

--- a/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
+++ b/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
@@ -875,7 +875,7 @@ class MagmadUtil(object):
         if self._init_system is None:
             self._init_system = self.detect_init_system()
 
-    def exec_command(self, command):
+    def exec_command(self, command: str) -> int:
         """Run a command remotely on magma_dev VM.
 
         Args:
@@ -893,7 +893,7 @@ class MagmadUtil(object):
             stderr=subprocess.DEVNULL,
         )
 
-    def exec_command_output(self, command):
+    def exec_command_output(self, command: str) -> str:
         """Run a command remotely on magma_dev VM.
 
         Args:
@@ -909,15 +909,18 @@ class MagmadUtil(object):
             shell=False,
         ).decode("utf-8")
 
-    def exec_command_capture_output(self, command):
+    def exec_command_capture_output(self, command: str) -> subprocess.CompletedProcess:
         """Run a command remotely on magma_dev VM.
+
+        Unlike `exec_command_output`, this method does not cause an exception
+        if the command returns a non-zero error code.
 
         Args:
             command: command (str) to be executed on remote host
             e.g. 'sed -i \'s/config1/config2/g\' /etc/magma/mme.yml'
 
         Returns:
-            output of command execution
+            Output  of command execution as instance of subprocess.CompletedProcess
         """
         param_list = shlex.split(self._command.format(**self._credentials, command=f'"{command}"'))
         return subprocess.run(
@@ -935,10 +938,10 @@ class MagmadUtil(object):
             "systemctl is-active magma@magmad",
         )
         docker_magmad_running = \
-            self._is_installed("docker") & \
+            self._is_installed("docker") and \
             (res_docker.stdout.decode("utf-8").strip('\n') == 'magmad')
         systemd_magmad_running = \
-            self._is_installed("systemctl") & \
+            self._is_installed("systemctl") and \
             (res_systemd.stdout.decode("utf-8").strip('\n') == 'active')
 
         if systemd_magmad_running:

--- a/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
+++ b/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
@@ -931,29 +931,24 @@ class MagmadUtil(object):
 
     def detect_init_system(self) -> InitMode:
         """Detect whether services are running with Docker or systemd."""
-        res_docker = self.exec_command_capture_output(
-            "docker ps --filter 'name=magmad' --format '{{.Names}}'",
-        )
-        res_systemd = self.exec_command_capture_output(
-            "systemctl is-active magma@magmad",
-        )
-        docker_magmad_running = \
-            self._is_installed("docker") and \
-            (res_docker.stdout.decode("utf-8").strip('\n') == 'magmad')
-        systemd_magmad_running = \
-            self._is_installed("systemctl") and \
-            (res_systemd.stdout.decode("utf-8").strip('\n') == 'active')
+        if self._is_installed("systemctl"):
+            res_systemd = self.exec_command_capture_output(
+                "systemctl is-active magma@magmad",
+            ).stdout.decode("utf-8").strip('\n')
+            if res_systemd == 'active':
+                # default to systemd if docker and systemd are running - needed by feg integ tests
+                return InitMode.SYSTEMD
 
-        if systemd_magmad_running:
-            # default to systemd if docker and systemd are running - needed by feg integ tests
-            return InitMode.SYSTEMD
-        elif docker_magmad_running:
-            return InitMode.DOCKER
-        else:
-            raise RuntimeError(
-                "Magmad is not running, you have to start magmad "
-                "either in Docker or systemd",
-            )
+        if self._is_installed("docker"):
+            res_docker = self.exec_command_capture_output(
+                "docker ps --filter 'name=magmad' --format '{{.Names}}'",
+            ).stdout.decode("utf-8").strip('\n')
+            if res_docker == 'magmad':
+                return InitMode.DOCKER
+        raise RuntimeError(
+            "Magmad is not running, you have to start magmad "
+            "either in Docker or systemd",
+        )
 
     def _is_installed(self, cmd):
         """Check if a command is installed on the system."""

--- a/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
+++ b/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
@@ -858,11 +858,10 @@ class MagmadUtil(object):
         """
         self._magmad_client = magmad_client
 
-        self._data = {
+        self._credentials = {
             "user": "vagrant",
             "host": "192.168.60.142",
             "password": "vagrant",
-            "command": "test",
         }
 
         self._command = (
@@ -886,9 +885,7 @@ class MagmadUtil(object):
         Returns:
             status of command execution
         """
-        data = self._data
-        data["command"] = '"' + command + '"'
-        param_list = shlex.split(self._command.format(**data))
+        param_list = shlex.split(self._command.format(**self._credentials, command=f'"{command}"'))
         return subprocess.call(
             param_list,
             shell=False,
@@ -906,35 +903,39 @@ class MagmadUtil(object):
         Returns:
             output of command execution
         """
-        data = self._data
-        data["command"] = '"' + command + '"'
-        param_list = shlex.split(self._command.format(**data))
+        param_list = shlex.split(self._command.format(**self._credentials, command=f'"{command}"'))
         return subprocess.check_output(
             param_list,
             shell=False,
         ).decode("utf-8")
 
+    def exec_command_capture_output(self, command):
+        """Run a command remotely on magma_dev VM.
+
+        Args:
+            command: command (str) to be executed on remote host
+            e.g. 'sed -i \'s/config1/config2/g\' /etc/magma/mme.yml'
+
+        Returns:
+            output of command execution
+        """
+        param_list = shlex.split(self._command.format(**self._credentials, command=f'"{command}"'))
+        return subprocess.run(
+            param_list,
+            shell=False,
+            capture_output=True,
+        )
+
     def detect_init_system(self) -> InitMode:
         """Detect whether services are running with Docker or systemd."""
-        def _is_installed(cmd):
-            is_installed = self.exec_command(f"type {cmd} >/dev/null 2>&1") == 0
-            if not is_installed:
-                logging.info(f"{cmd} is not installed")
-            return is_installed
-
-        if _is_installed("docker"):
-            docker_magmad_running = self.exec_command_output(
-                "docker ps --filter 'name=magmad' --format '{{.Names}}'",
-            ).strip() == "magmad"
-        else:
-            docker_magmad_running = False
-
-        if _is_installed("systemctl"):
-            systemd_magmad_running = self.exec_command_output(
-                "systemctl is-active magma@magmad",
-            ).strip() == "active"
-        else:
-            systemd_magmad_running = False
+        res_docker = self.exec_command_capture_output(
+            "docker ps --filter 'name=magmad' --format '{{.Names}}'",
+        )
+        res_systemd = self.exec_command_capture_output(
+            "systemctl is-active magma@magmad",
+        )
+        docker_magmad_running = res_docker.stdout.decode("utf-8").strip('\n') == 'magmad'
+        systemd_magmad_running = res_systemd.stdout.decode("utf-8").strip('\n') == 'active'
 
         if systemd_magmad_running:
             # default to systemd if docker and systemd are running - needed by feg integ tests

--- a/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
+++ b/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
@@ -909,51 +909,25 @@ class MagmadUtil(object):
             shell=False,
         ).decode("utf-8")
 
-    def exec_command_capture_output(self, command: str) -> subprocess.CompletedProcess:
-        """Run a command remotely on magma_dev VM.
-
-        Unlike `exec_command_output`, this method does not cause an exception
-        if the command returns a non-zero error code.
-
-        Args:
-            command: command (str) to be executed on remote host
-            e.g. 'sed -i \'s/config1/config2/g\' /etc/magma/mme.yml'
-
-        Returns:
-            Output  of command execution as instance of subprocess.CompletedProcess
-        """
-        param_list = shlex.split(self._command.format(**self._credentials, command=f'"{command}"'))
-        return subprocess.run(
-            param_list,
-            shell=False,
-            capture_output=True,
-        )
-
     def detect_init_system(self) -> InitMode:
         """Detect whether services are running with Docker or systemd."""
-        res_docker = self.exec_command_capture_output(
-            "docker ps --filter 'name=magmad' --format '{{.Names}}'",
+        if self._is_installed("systemctl"):
+            ret = self.exec_command_output(
+                "systemctl is-active magma@magmad",
+            ).strip('\n')
+            if ret == 'active':
+                # default to systemd if docker and systemd are running - needed by feg integ tests
+                return InitMode.SYSTEMD
+        if self._is_installed("docker"):
+            ret = self.exec_command_output(
+                "docker ps --filter 'name=magmad' --format '{{.Names}}'",
+            ).strip('\n')
+            if ret == 'magmad':
+                return InitMode.DOCKER
+        raise RuntimeError(
+            "Magmad is not running, you have to start magmad "
+            "either in Docker or systemd",
         )
-        res_systemd = self.exec_command_capture_output(
-            "systemctl is-active magma@magmad",
-        )
-        docker_magmad_running = \
-            self._is_installed("docker") and \
-            (res_docker.stdout.decode("utf-8").strip('\n') == 'magmad')
-        systemd_magmad_running = \
-            self._is_installed("systemctl") and \
-            (res_systemd.stdout.decode("utf-8").strip('\n') == 'active')
-
-        if systemd_magmad_running:
-            # default to systemd if docker and systemd are running - needed by feg integ tests
-            return InitMode.SYSTEMD
-        elif docker_magmad_running:
-            return InitMode.DOCKER
-        else:
-            raise RuntimeError(
-                "Magmad is not running, you have to start magmad "
-                "either in Docker or systemd",
-            )
 
     def _is_installed(self, cmd):
         """Check if a command is installed on the system."""

--- a/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
+++ b/lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py
@@ -909,25 +909,51 @@ class MagmadUtil(object):
             shell=False,
         ).decode("utf-8")
 
+    def exec_command_capture_output(self, command: str) -> subprocess.CompletedProcess:
+        """Run a command remotely on magma_dev VM.
+
+        Unlike `exec_command_output`, this method does not raise an exception
+        if the command returns a non-zero error code.
+
+        Args:
+            command: command (str) to be executed on remote host
+            e.g. 'sed -i \'s/config1/config2/g\' /etc/magma/mme.yml'
+
+        Returns:
+            Output  of command execution as instance of subprocess.CompletedProcess
+        """
+        param_list = shlex.split(self._command.format(**self._credentials, command=f'"{command}"'))
+        return subprocess.run(
+            param_list,
+            shell=False,
+            capture_output=True,
+        )
+
     def detect_init_system(self) -> InitMode:
         """Detect whether services are running with Docker or systemd."""
-        if self._is_installed("systemctl"):
-            ret = self.exec_command_output(
-                "systemctl is-active magma@magmad",
-            ).strip('\n')
-            if ret == 'active':
-                # default to systemd if docker and systemd are running - needed by feg integ tests
-                return InitMode.SYSTEMD
-        if self._is_installed("docker"):
-            ret = self.exec_command_output(
-                "docker ps --filter 'name=magmad' --format '{{.Names}}'",
-            ).strip('\n')
-            if ret == 'magmad':
-                return InitMode.DOCKER
-        raise RuntimeError(
-            "Magmad is not running, you have to start magmad "
-            "either in Docker or systemd",
+        res_docker = self.exec_command_capture_output(
+            "docker ps --filter 'name=magmad' --format '{{.Names}}'",
         )
+        res_systemd = self.exec_command_capture_output(
+            "systemctl is-active magma@magmad",
+        )
+        docker_magmad_running = \
+            self._is_installed("docker") and \
+            (res_docker.stdout.decode("utf-8").strip('\n') == 'magmad')
+        systemd_magmad_running = \
+            self._is_installed("systemctl") and \
+            (res_systemd.stdout.decode("utf-8").strip('\n') == 'active')
+
+        if systemd_magmad_running:
+            # default to systemd if docker and systemd are running - needed by feg integ tests
+            return InitMode.SYSTEMD
+        elif docker_magmad_running:
+            return InitMode.DOCKER
+        else:
+            raise RuntimeError(
+                "Magmad is not running, you have to start magmad "
+                "either in Docker or systemd",
+            )
 
     def _is_installed(self, cmd):
         """Check if a command is installed on the system."""

